### PR TITLE
(WIP) osVoice decomp tracking

### PR DESCRIPTION
**Pull request by someone2639**
Important things to review:
-the Decompollaborate/mm repo has some of these decompiled in some form, which I adapted to this branch; however, defines such as `CONT_ERR_CONTRFAIL` are not in this repo by default, while they do show up in the `mm` repo

**Very** important things to review
-I wrote an essay in src/voice/voicecountsyllables.c and I have no idea if it's correct or useful

Will update description as I go

### CODE REVIEW

File: include/controller.h [hunk 1]
```diff
@@ -0,0 +1,163 @@
+#ifndef _CONTROLLER_H_
```
Comment (Mr-Wiseguy) [inline include/controller.h:1]: We have a controller.h in the repo already in src/io, is this one needed? This one seems to have some game-specific variables/structs and custom button defines that are equivalent to the standard ones.

File: include/PR/os_voice.h [hunk 2]
```diff
@@ -0,0 +1,99 @@
+#ifndef _OS_VOICE_H_
```
Comment (Mr-Wiseguy) [inline include/PR/os_voice.h:1]: Go ahead and re-add the notice at the top of this file, we'll be reverting that change to all other files shortly

File: include/PR/os.h [hunk 3]
```diff
@@ -1,795 +1,794 @@
-
```
Comment (Mr-Wiseguy) [inline include/PR/os.h:1]: If you rebase on current main, these line endings will have already been fixed

**Review by Mr-Wiseguy — CHANGES_REQUESTED** [review]

File: src/voice/voicecountsyllables.c [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+        } else if (IS_ENDING_GLYPH(sjis)) {
+            // ん, っ, and ー all have one semisyllable
+            *syllable += 1;
+        } else {
+            // all other cases add 2 semisyllables
+            *syllable += 2;
+        }
+        old = sjis;
+    }
+}
```
Comment (EllipticEllipsis) [inline src/voice/voicecountsyllables.c:78]:
```suggestion
}

```

File: src/voice/voiceinit.c [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+#include "../io/controller.h"
+#include "PR/os_cont.h"
+#include "PR/os_voice.h"
+
+s32 osVoiceInit(OSMesgQueue* mq, OSVoiceHandle* handle, int channel) {
+    s32 ret;
+    s32 i;
+    u8 stat = 0;
+    u8 buf[4];
+    static u8 cmd[] = {0x1E, 0x6E, 0x08, 0x56, 0x03};
```
Comment (EllipticEllipsis) [inline src/voice/voiceinit.c:11]: Presumably there will be macros associated to these in future?

File: src/voice/voicecontrolgain.c [hunk 6]
```diff
@@ -0,0 +1,24 @@
+#include "PR/os_internal.h"
+#include "../io/controller.h"
+#include "PR/os_voice.h"
+
+s32 __osVoiceSetADConverter(OSMesgQueue*, int ch, u8 cmd);
+
+#define ACOUNT(a) (sizeof((a)) / sizeof((a[0])))
```
Comment (EllipticEllipsis) [inline src/voice/voicecontrolgain.c:7]: I have to admit I'm somewhat baffled by the presence of this macro here, when it doesn't seem to be used at all in this file.

**Review by EllipticEllipsis — COMMENTED** [review]
It's tricky to walk the line between respecting the SDK names and accurately documenting what this is actually doing. I've made a few linguistically-inclined comments, but I'm not sure how many are actually going to be possible to act on.

File: src/io/controller.h [hunk 7]
```diff
@@ -144,6 +153,17 @@ typedef struct
 #define DIR_STATUS_UNKNOWN 1
 #define DIR_STATUS_OCCUPIED 2
 
+/* Controller error number */
```
Comment (Mr-Wiseguy) [inline src/io/controller.h:156]: These defines are already present in os_cont.h

File: src/io/controller.h [hunk 8]
```diff
@@ -47,6 +47,15 @@ typedef struct
...
 } __OSContRequesFormatShort;
 
+typedef struct {
+    /* 0x00 */ u8 txsize;
+    /* 0x01 */ u8 rxsize;
+    /* 0x02 */ u8 poll;
+    /* 0x03 */ u8 typeh;
+    /* 0x04 */ u8 typel;
+    /* 0x05 */ u8 status;
+} __OSContRequestHeaderAligned; // size = 0x6
```
Comment (Mr-Wiseguy) [inline src/io/controller.h:57]: This struct should be named "__OSContRequesFormatShort" and the "poll" member should be named "cmd"

File: src/voice/voicecontread36.c [hunk 9]
```diff
@@ -0,0 +1,69 @@
...
+            __osContLastCmd = 9;
+            __osPfsLastChannel = channel;
+
+            for (i = 0; i < channel; i++) {
+                *ptr++ = 0;
+            }
+
+            __osPfsPifRam.pifstatus = CONT_CMD_READ_BUTTON;
+
+            ptr[0] = 0xFF;
```
Comment (Mr-Wiseguy) [inline src/voice/voicecontread36.c:29]: These correspond to struct offsets to structs in controller_voice.h, which you can find with libelf.py. Take a look at this scratch to see how they'd get used: https://decomp.me/scratch/PTDM7

**Review by Mr-Wiseguy — CHANGES_REQUESTED** [review]

**Comment by someone2639** [comment]
just kidding _now_ everything is ready

File: src/voice/voicesetword.c [hunk 10]
```diff
@@ -0,0 +1,46 @@
...
+    if (stat & 2) {
+        return 0xF;
+    }
+    for (k = 0; word[k] != 0; k += 2) {
+        ;
+    }
+    // if (k >= 0x22) {
+    //     return 0xE;
+    // }
+    bzero(buf, 40);
```
Comment (Mr-Wiseguy) [inline src/voice/voicesetword.c:22]: Same deal here with the constants used for buf as in voicemaskdictionary.c

File: src/voice/voicesetword.c [hunk 11]
```diff
@@ -0,0 +1,46 @@
...
+    u8 buf[40];
+
+    ERRCK(__osVoiceGetStatus(hd->__mq, hd->__channel, &stat));
+    if (stat & 2) {
+        return 0xF;
+    }
+    for (k = 0; word[k] != 0; k += 2) {
+        ;
+    }
+    // if (k >= 0x22) {
```
Comment (Mr-Wiseguy) [inline src/voice/voicesetword.c:19]: Commented code can be removed.

File: src/voice/voicesetword.c [hunk 12]
```diff
@@ -0,0 +1,46 @@
...
+    buf[39 - j - 5] = 3;
+
+    if (k >= 15) {
+        ERRCK(__osVoiceContWrite20(hd->__mq, hd->__channel, 0, buf));
+    }
+    ERRCK(__osVoiceContWrite20(hd->__mq, hd->__channel, 0, buf + 20));
+    ret = __osVoiceCheckResult(hd, &stat);
+    if (ret != 0) {
+        if (ret & 0x100) {
+            ret = 0xD;
```
Comment (Mr-Wiseguy) [inline src/voice/voicesetword.c:38]: These return values are one of CONT_ERR_x in os_cont.h

File: src/voice/voicegetstatus.c [hunk 13]
```diff
@@ -0,0 +1,61 @@
...
+
+    do {
+        if (ret != CONT_ERR_CONTRFAIL) {
+            __osContPifRam.pifstatus = CONT_CMD_EXE;
+
+            for (i = 0; i < port; i++, *ptr++ = 0) {
+                ;
+            }
+
+            *ptr++ = 1;
```
Comment (Mr-Wiseguy) [inline src/voice/voicegetstatus.c:23]: These two lines are another pair of TX and RX sizes.

File: src/voice/voicegetreaddata.c [hunk 14]
```diff
@@ -0,0 +1,94 @@
...
+                return CONT_ERR_VOICE_NO_RESPONSE;
+            }
+
+            break;
+        default:
+            return CONT_ERR_INVALID;
+    }
+    hd->__mode = 0;
+    return ret;
+}
```
Comment (Mr-Wiseguy) [inline src/voice/voicegetreaddata.c:94]: No newline at end of file

File: src/voice/voicecontread36.c [hunk 15]
```diff
@@ -0,0 +1,73 @@
...
+
+        __osSiRawStartDma(OS_WRITE, &__osPfsPifRam);
+        osRecvMesg(mq, NULL, OS_MESG_BLOCK);
+        __osSiRawStartDma(OS_READ, &__osPfsPifRam);
+        osRecvMesg(mq, NULL, OS_MESG_BLOCK);
+
+        ret = (READ36FORMAT(ptr)->rxsize & 0xC0) >> 4;
+
+        if (ret == 0) {
+            if (__osVoiceContDataCrc(&READ36FORMAT(ptr)->data, 36) != READ36FORMAT(ptr)->datacrc) {
```
Comment (Mr-Wiseguy) [inline src/voice/voicecontread36.c:54]: The 36 here can be an `ARRLEN(READ36FORMAT(ptr)->data)`

File: src/voice/voicecontread36.c [hunk 16]
```diff
@@ -0,0 +1,73 @@
...
+
+        READ36FORMAT(ptr)->addrh = address >> 3;
+        READ36FORMAT(ptr)->addrl = __osContAddressCrc(address) | (address << 5);
+
+        __osSiRawStartDma(OS_WRITE, &__osPfsPifRam);
+        osRecvMesg(mq, NULL, OS_MESG_BLOCK);
+        __osSiRawStartDma(OS_READ, &__osPfsPifRam);
+        osRecvMesg(mq, NULL, OS_MESG_BLOCK);
+
+        ret = (READ36FORMAT(ptr)->rxsize & 0xC0) >> 4;
```
Comment (Mr-Wiseguy) [inline src/voice/voicecontread36.c:51]: This is the `CHNL_ERR` macro from controller.h

File: src/voice/voicecontread36.c [hunk 17]
```diff
@@ -0,0 +1,73 @@
...
+        if (ret == 0) {
+            if (__osVoiceContDataCrc(&READ36FORMAT(ptr)->data, 36) != READ36FORMAT(ptr)->datacrc) {
+                ret = __osVoiceGetStatus(mq, channel, &status);
+                if (ret != 0) {
+                    break;
+                }
+
+                ret = CONT_ERR_CONTRFAIL;
+            } else {
+                bcopy(&READ36FORMAT(ptr)->data, buffer, 36);
```
Comment (Mr-Wiseguy) [inline src/voice/voicecontread36.c:62]: Same as the ARRLEN comment above.

File: src/voice/voicecontrolgain.c [hunk 18]
```diff
@@ -0,0 +1,24 @@
+#include "PR/os_internal.h"
+#include "io/controller.h"
+#include "PR/os_voice.h"
+
+s32 __osVoiceSetADConverter(OSMesgQueue*, int ch, u8 cmd);
+
+#define ACOUNT(a) (sizeof((a)) / sizeof((a[0])))
```
Comment (Mr-Wiseguy) [inline src/voice/voicecontrolgain.c:7]: This macro is unused and can be removed.

File: src/voice/voicegetstatus.c [hunk 19]
```diff
@@ -0,0 +1,61 @@
...
+            osRecvMesg(mq, NULL, OS_MESG_BLOCK);
+        }
+        ret = __osSiRawStartDma(OS_READ, &__osContPifRam);
+        osRecvMesg(mq, NULL, OS_MESG_BLOCK);
+
+        ptr = (u8*)&__osContPifRam.ramarray + port;
+
+        header = *((__OSContRequesFormatShort*)ptr);
+
+        ret = (u8)((header.rxsize & 0xC0) >> 4);
```
Comment (Mr-Wiseguy) [inline src/voice/voicegetstatus.c:40]: This is the `CHNL_ERR` macro in controller.h

**Review by Mr-Wiseguy — CHANGES_REQUESTED** [review]
Mostly formatting or symbolization changes.

**Comment by someone2639** [comment]
Comments addressed

**Review by Mr-Wiseguy — APPROVED** [review]
Lgtm

**Pull request merged**